### DISCUSSION TITLE
Stopping Bonsai From Command Line

### DISCUSSION THREAD

**Discussion by spencerblackwood** [bonsai-rx/bonsai#893] [Q&A] [answered]
Is there a way to stop a Bonsai workflow using the command line interface?

I would like to automate data collection by starting and stopping Bonsai workflows with different parameters through the command line interface. Launching the Bonsai workflows and setting externalized properties is fairly straightforward. However, you can't stop the workflow, and I don't know how to create a self-terminating workflow with continuous data source nodes.

Using im/taskkill Bonsai.exe seems like a bad idea, so does anyone have any ideas?

-Spencer

**Comment by bruno-f-cruz** [comment]
Currently, There is no implementation that allows sending a stop command to Bonsai using CLI. Using task-kill might indeed be a bad idea since it will abruptly abort some processes such as saving videos, csv, etc...
We might be able to help you come up with an alternative solution if you have an idea on what condition you would like to use to close bonsai.
Cheers!

**Reply by spencerblackwood** [reply]
We are doing long-term EEG in mice (multiple weeks), and it would be helpful to have the computer start/stop the recording at specified intervals then immediately begin processing and data backup on the previous recording. Of course it can be done by hand, but having exact recording lengths and automated processing/backup would be neat.

**Reply by bruno-f-cruz** [reply]
This sounds doable. You can stop an workflow after a pre-determined period of time by connecting a `Timer` with a given `DueTime` property to a `WorkflowOutput` at the highest level of the workflow. If you want something a bit more concrete let me know exactly what would work for you and I will throw in an example!

**Comment by jsiegle** [comment, marked as answer]
What about inserting an OSC operator that ends the workflow when it receives a message? Then you could launch another workflow from the command line that sends an OSC message before terminating.

**Reply by spencerblackwood** [reply]
Any idea how I can stop the workflow from within Bonsai after receiving the OSC message? I am recording EEG using open ephys boards (hi Josh!), and I don't think the workflow will stop as long as the RHDEvalBoard node is running.

**Reply by bruno-f-cruz** [reply]
Similarly to what I said before, once a branch connected to a `WorkflowOutput` node is closed, it will result on that workflow to close. You can exploit this behavior to have any kind of stream ending the workflow. Following @jsiegle suggestion, a simple `OscMessageEvent`-> `Take` (closes the branch) -> `WorkflowOutput` will close the workflow.

**Reply by spencerblackwood** [reply]
Yes, the ReceiveMessage -> Take  -> WorkflowOutput pattern does exactly what I want. Thanks for your help!
